packages/runner/src/run.ts

```diff
@@ -150,11 +152,11 @@ export async function callSuiteHook<T extends keyof SuiteHooks>(
   }
 
   async function runHook(hook: Function) {
-    return getBeforeHookCleanupCallback(
+    return limitMaxConcurrency(async () => getBeforeHookCleanupCallback(
```

Comment:
This needs to be careful not to include `beforeEach/afterEach`. "each" hooks are called inside `runTest`, which already run inside `limitMaxConcurrency`, so they can potentially deadlock. For the test case, we should verify these cases won't get stuck:

```js
// --max-concurrency=1
beforeEach(async () => {})
test("repro", async () => {})
afterEach(async () => {})
```

```js
// --max-concurrency=1
beforeAll(async () => {})
test("repro", async () => {})
afterAll(async () => {})
```